lib/streamlit/testing/v1/local_script_runner.py

```diff
@@ -0,0 +1,144 @@
+        )
+
+        # Accumulates all ScriptRunnerEvents emitted by us.
+        self.events: list[ScriptRunnerEvent] = []
+        self.event_data: list[Any] = []
+
+        def record_event(
+            sender: ScriptRunner | None, event: ScriptRunnerEvent, **kwargs
+        ) -> None:
+            # Assert that we're not getting unexpected `sender` params
+            # from ScriptRunner.on_event
+            assert (
+                sender is None or sender == self
+            ), "Unexpected ScriptRunnerEvent sender!"
+
+            self.events.append(event)
+            self.event_data.append(kwargs)
+
+            # Send ENQUEUE_FORWARD_MSGs to our queue
+            if event == ScriptRunnerEvent.ENQUEUE_FORWARD_MSG:
+                forward_msg = kwargs["forward_msg"]
+                self.forward_msg_queue.enqueue(forward_msg)
+
+        self.on_event.connect(record_event, weak=False)
+
+    def join(self) -> None:
+        """Wait for the script thread to finish, if it is running."""
+        if self._script_thread is not None:
+            self._script_thread.join()
+
+    def forward_msgs(self) -> list[ForwardMsg]:
+        """Return all messages in our ForwardMsgQueue."""
+        return self.forward_msg_queue._queue
+
+    def run(
+        self,
+        widget_state: WidgetStates | None = None,
+        query_params=None,
+        timeout: float = 3,
+    ) -> ElementTree:
+        """Run the script, and parse the output messages for querying
+        and interaction.
+
+        Timeout is in seconds.
+        """
+        query_string = ""
+        if query_params:
+            query_string = parse.urlencode(query_params, doseq=True)
+
+        rerun_data = RerunData(widget_states=widget_state, query_string=query_string)
+        self.request_rerun(rerun_data)
+        if not self._script_thread:
+            self.start()
+        require_widgets_deltas(self, timeout)
+
+        tree = parse_tree_from_messages(self.forward_msgs())
+        return tree
+
+    def script_stopped(self) -> bool:
+        for e in self.events:
+            if e in (
+                ScriptRunnerEvent.SCRIPT_STOPPED_FOR_RERUN,
+                ScriptRunnerEvent.SCRIPT_STOPPED_WITH_COMPILE_ERROR,
+                ScriptRunnerEvent.SCRIPT_STOPPED_WITH_SUCCESS,
+            ):
+                return True
+        return False
+
+
+def require_widgets_deltas(runner: LocalScriptRunner, timeout: float = 3) -> None:
+    """Wait for the given ScriptRunner to emit a completion event. If the timeout
+    is reached, the runner will be shutdown and an error will be thrown.
+    """
+
+    t0 = time.time()
+    while time.time() - t0 < timeout:
+        time.sleep(0.1)
+        if runner.script_stopped():
+            return
+
+    # If we get here, the runner hasn't yet completed before our
+    # timeout. Create an error string for debugging.
+    err_string = f"require_widgets_deltas() timed out after {timeout}s)"
```

Comment:
This is the error string returned the the end user on a timeout, they won't know about require_widgets_deltas - should it be something like `AppTest run timed out after {timeout}s`? Or caught further up the stack and given a friendly error? Also has an extra trailing parenthesis in the string